File: problems-adapt/0101-0200/0167_pair-sum-in-order/solution_two_pointers.cpp
class Solution {
  public:
    vector<int> pairSumInOrder(vector<int> &nums, int target) {
        int left = 0, right = (int)nums.size() - 1;
        while (left < right) {
            int total = nums[left] + nums[right];
            if (total == target)
                // 1-based indices as the problem expects.
                return {left + 1, right + 1};
            if (total < target)
                // Too small: pairing nums[left] with anything smaller than
                // nums[right] only lowers the sum — retire the left value.
                ++left;
            else
                // Too large: retire the right value symmetrically.
                --right;
        }
        // Unreachable under the uniqueness promise; keeps the function total.
        return {};
    }
};

File: problems-originals/0101-0200/0167_two-sum-ii-input-array-is-sorted/solution_two_pointers.go
func twoSum(numbers []int, target int) []int {
	left, right := 0, len(numbers)-1
	for left < right {
		total := numbers[left] + numbers[right]
		switch {
		case total == target:
			// 1-based indices as the problem expects.
			return []int{left + 1, right + 1}
		case total < target:
			// Too small: pairing numbers[left] with anything smaller than
			// numbers[right] only lowers the sum — retire the left value.
			left++
		default:
			// Too large: retire the right value symmetrically.
			right--
		}
	}
	// Unreachable under the uniqueness promise; keeps the function total.
	return nil
}
